chapters/chap04-sec03.tex

```diff
@@ -48,7 +48,26 @@ \subsubsection{Zero-knowledge proof generation}
 
 See \cref{appendix:sca-attacks:proof-generation} for a discussion of related security concerns.
 
-\subsection{Contract performance}\label{implementation:efficiency:contract-performance}
+\subsection{Zero-knowledge proof verification (on-chain)}
+
+Verification of the joinsplit statement via a zero-knowledge proof represents a significant computation, which must be carried out on-chain (by the \mixer~contract) for each valid \zeth~transaction. As described in \cref{instantiation:zksnark}, this verification cost increases linearly with the number of primary inputs to the statement - a scalar multiplication of a group element and a group addition operation must be performed for each primary input. A technique presented in \cite[Section 4.5.1]{GGPR13} can be applied to reduce this linear cost.
+
+Given a relation $\REL$ with language $\LANG$, and a collision resistant hash function $H : \LANG \to \FFx{\rCURVE}$, let $\REL^\prime = \LANG^\prime \times \REL$ be a new relation, with corresponding language $\LANG^\prime \subset \FFx{\rCURVE}$, such that $(\priminputs^\prime, \auxinputs^\prime) \in \REL^\prime$ iff $\priminputs^\prime = H(\priminputs)$, where $\auxinputs^\prime = (\priminputs, \auxinputs) \in \REL$. To verify that $\priminputs \in \LANG$, a verifier can compute $H(\priminputs)$ and check that $H(\priminputs) \in \LANG^\prime$. (By construction, if $H(\priminputs) \in \LANG^\prime$, there exists $(\priminputs_0, \auxinputs) \in \REL$, i.e. $\priminputs_0 \in LANG$ with $H(\priminputs_0) == H(\priminputs)$. By the collision-resistance of $H$ we have $\priminputs_0 == \priminputs$ with overwhelming probability.
```

Comment:
I wanted to give a basic formulation in terms of the concepts from the preliminaries chapter, but this may not be particularly clear.  It mght make more sense to put the intuitive description (below) before this formulation, or potentially just remove this formulation and refer the reader to the (clearer) description in the original paper.   @AntoineRondelet let me know what you think.

Reply:
Thanks @dtebbs - I've just assigned me as a reviewer, will have a look

Reply:
I find your formulation to be pretty clear (at least - the version in https://github.com/clearmatics/zeth-specifications/pull/15/commits/58fb183a073ffa39d2dac789360b29e684ccb0d0 :) )